modules/bibdk_heimdal/includes/bibdk_heimdal.pages.inc

```diff
@@ -0,0 +1,184 @@
+    // for another login attempt
+    $message = bibdk_heimdal_pages_welcome_message();
+    drupal_set_message($message);
+    $link = bibdk_heimdal_pages_get_login_link();
+    drupal_goto($link);
+  }
+
+  // user was not created on openuserinfo - set a  message and return to front
+  drupal_set_message(bibdk_heimdal_pages_something_went_wrong_message(), 'error');
+  drupal_goto('<front>');
+}
+
+/**
+ * Create a user on openuserinfo - if user doesn't already exist
+ * @param string $mail
+ *   Username
+ */
+function bibdk_heimdal_pages_create_bibdk_user($mail) {
+  if (!BibdkUser::instance()->verify($mail)) {
+    // user does not exist - create
+    // make a random password
+    $pass = user_password();
+    return bibdk_provider_user_create_account($mail, $pass);
+  }
+  return TRUE;
+}
+
+function bibdk_heimdal_pages_something_went_wrong_message(){
+  $message = t('Something went wrong - please try again. Contact administrator if problem persists',
+    array(), array('context'=>'heimdal'));
+  return $message;
+}
+
+/**
+ * Get a message
+ * @return null|string
+ */
+function bibdk_heimdal_pages_welcome_message() {
+  $message = t('welcome_to_single_sign_on', array(), array('context'=>'heimdal'));
+  return $message;
+}
+
+/**
+ * Get a message
+ * @return null|string
+ */
+function bibdk_heimdal_pages_no_culr_message(){
+  $message = t('bibdk_heimdal_no_culr_helptext', array(), array('context' => 'heimdal'));
+  return $message;
+}
+
+
+
+/**
+ * check response from createAccount action on culr
+ * @param $response
+ * @return bool
+ */
+function bibdk_heimcdal_pages_check_insert_response($response){
+  // @TODO implement
+  return TRUE;
+}
+
+/**
+ * Check if given time is more than a day old
+ * @param int $time
+ * @return bool
+ */
+function bibdk_heimdal_check_timestamp(int $time){
+  $now = time();
+  // 86400 seconds = one day
+  if($now - $time > 86400){
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+ * Let openuserinfo do an insert(createAccount) on culr.
+ * @param $row
+ * @return array
+ */
+function bibdk_heimdal_do_insert_culr($row){
+  $agencies = unserialize($row['libraries']);
+  if(empty($agencies)){
+    $agencies = array();
+  }
+  $response = array();
+  $mail = $row['email'];
+  foreach ($agencies as $agency) {
+    $borrowerId = $agency['userId'];
+    $uidType = $agency['userIdType'];
+    $response[$agency['agencyId']][] = BibdkUser::instance()->insertCulr($mail, $borrowerId, $uidType);
+  }
+  return $response;
+}
+
+/**
+ * Get a login link for heimdal access platform
+ * @return string
+ */
+function bibdk_heimdal_pages_get_login_link() {
+  $config = variable_get('ding_wayf', NULL);
+  $token = variable_get('heimdal_access_token');
+  $heimdal_link = $config['hosturl'] . $config['loginpath'] . '?token=' . $token . '&destination=heimdal/login';
+  return $heimdal_link;
+}
+
+/**
+ * Get a message
+ * @return null|string
+ */
+function bibdk_heimdal_exception_message(){
```

Comment:
This could be joined with the bibdk_heimdal_pages_something_went_wrong_message function.